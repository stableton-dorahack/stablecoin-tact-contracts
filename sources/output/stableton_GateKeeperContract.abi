{"name":"GateKeeperContract","types":[{"name":"StateInit","header":null,"fields":[{"name":"code","type":{"kind":"simple","type":"cell","optional":false}},{"name":"data","type":{"kind":"simple","type":"cell","optional":false}}]},{"name":"Context","header":null,"fields":[{"name":"bounced","type":{"kind":"simple","type":"bool","optional":false}},{"name":"sender","type":{"kind":"simple","type":"address","optional":false}},{"name":"value","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"raw","type":{"kind":"simple","type":"slice","optional":false}}]},{"name":"SendParameters","header":null,"fields":[{"name":"bounce","type":{"kind":"simple","type":"bool","optional":false}},{"name":"to","type":{"kind":"simple","type":"address","optional":false}},{"name":"value","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"mode","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"body","type":{"kind":"simple","type":"cell","optional":true}},{"name":"code","type":{"kind":"simple","type":"cell","optional":true}},{"name":"data","type":{"kind":"simple","type":"cell","optional":true}}]},{"name":"Deploy","header":2490013878,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}}]},{"name":"DeployOk","header":2952335191,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}}]},{"name":"ChangeOwner","header":256331011,"fields":[{"name":"newOwner","type":{"kind":"simple","type":"address","optional":false}}]},{"name":"PoolSettings","header":null,"fields":[{"name":"liquidationRatio","type":{"kind":"simple","type":"uint","optional":false,"format":32}},{"name":"stabilityFeeRate","type":{"kind":"simple","type":"uint","optional":false,"format":32}},{"name":"closeFactorBps","type":{"kind":"simple","type":"uint","optional":false,"format":32}},{"name":"liquidatorIncentiveBps","type":{"kind":"simple","type":"uint","optional":false,"format":32}}]},{"name":"DebtRate","header":null,"fields":[{"name":"debtAccumulatedRate","type":{"kind":"simple","type":"uint","optional":false,"format":32}},{"name":"lastAccumulationTime","type":{"kind":"simple","type":"uint","optional":false,"format":32}}]},{"name":"DepositCollateralMessage","header":3723304936,"fields":[{"name":"user","type":{"kind":"simple","type":"address","optional":false}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}},{"name":"settings","type":{"kind":"simple","type":"PoolSettings","optional":false}},{"name":"rate","type":{"kind":"simple","type":"DebtRate","optional":false}},{"name":"tonPriceWithSafetyMargin","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}}]},{"name":"WithdrawCollateralMessage","header":3410368520,"fields":[{"name":"user","type":{"kind":"simple","type":"address","optional":false}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}},{"name":"settings","type":{"kind":"simple","type":"PoolSettings","optional":false}},{"name":"rate","type":{"kind":"simple","type":"DebtRate","optional":false}},{"name":"tonPriceWithSafetyMargin","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}}]},{"name":"WithdrawStablecoinMessage","header":1815710139,"fields":[{"name":"user","type":{"kind":"simple","type":"address","optional":false}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":64}},{"name":"settings","type":{"kind":"simple","type":"PoolSettings","optional":false}},{"name":"rate","type":{"kind":"simple","type":"DebtRate","optional":false}},{"name":"tonPriceWithSafetyMargin","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}}]},{"name":"RepayStablecoinMessage","header":3203603975,"fields":[{"name":"user","type":{"kind":"simple","type":"address","optional":false}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":64}},{"name":"settings","type":{"kind":"simple","type":"PoolSettings","optional":false}},{"name":"rate","type":{"kind":"simple","type":"DebtRate","optional":false}},{"name":"tonPriceWithSafetyMargin","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}}]},{"name":"DoWithdrawCollateralMessage","header":3285005247,"fields":[{"name":"user","type":{"kind":"simple","type":"address","optional":false}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":64}}]},{"name":"IncreaseTotalStableMessage","header":2517198105,"fields":[{"name":"user","type":{"kind":"simple","type":"address","optional":false}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":64}}]},{"name":"DecreaseTotalStableMessage","header":3984666698,"fields":[{"name":"user","type":{"kind":"simple","type":"address","optional":false}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":64}},{"name":"fees","type":{"kind":"simple","type":"uint","optional":false,"format":64}}]},{"name":"MintFeesMessage","header":2638369821,"fields":[{"name":"to","type":{"kind":"simple","type":"address","optional":false}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}}]},{"name":"JettonData","header":null,"fields":[{"name":"totalSupply","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"mintable","type":{"kind":"simple","type":"bool","optional":false}},{"name":"owner","type":{"kind":"simple","type":"address","optional":false}},{"name":"content","type":{"kind":"simple","type":"cell","optional":true}},{"name":"walletCode","type":{"kind":"simple","type":"cell","optional":false}}]},{"name":"JettonWalletData","header":null,"fields":[{"name":"balance","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"owner","type":{"kind":"simple","type":"address","optional":false}},{"name":"master","type":{"kind":"simple","type":"address","optional":false}},{"name":"walletCode","type":{"kind":"simple","type":"cell","optional":false}}]},{"name":"SetUserPositionDependecyMessage","header":4030201083,"fields":[{"name":"stablecoinMasterAddress","type":{"kind":"simple","type":"address","optional":false}},{"name":"positionsManagerAddress","type":{"kind":"simple","type":"address","optional":false}}]},{"name":"SetPositionIdMessage","header":1245051344,"fields":[{"name":"user","type":{"kind":"simple","type":"address","optional":false}},{"name":"positionId","type":{"kind":"simple","type":"int","optional":false,"format":257}}]},{"name":"NewPositionIdMessage","header":225756565,"fields":[{"name":"user","type":{"kind":"simple","type":"address","optional":false}}]},{"name":"SetPositionAddressMessage","header":1571217046,"fields":[{"name":"user","type":{"kind":"simple","type":"address","optional":false}},{"name":"position","type":{"kind":"simple","type":"address","optional":false}}]},{"name":"MintMessage","header":2419415474,"fields":[{"name":"user","type":{"kind":"simple","type":"address","optional":false}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}}]},{"name":"RepayBurnMessage","header":2959459244,"fields":[{"name":"user","type":{"kind":"simple","type":"address","optional":false}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}},{"name":"fees","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}}]},{"name":"RepayBurnNotification","header":2355055729,"fields":[{"name":"user","type":{"kind":"simple","type":"address","optional":false}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}},{"name":"fees","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}}]},{"name":"StablecoinBurnedMessage","header":3929120156,"fields":[{"name":"user","type":{"kind":"simple","type":"address","optional":false}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}},{"name":"fees","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}}]},{"name":"TokenTransfer","header":260734629,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}},{"name":"destination","type":{"kind":"simple","type":"address","optional":false}},{"name":"responseDestination","type":{"kind":"simple","type":"address","optional":true}},{"name":"customPayload","type":{"kind":"simple","type":"cell","optional":true}},{"name":"forwardTonAmount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}},{"name":"forwardPayload","type":{"kind":"simple","type":"slice","optional":false,"format":"remainder"}}]},{"name":"TokenTransferInternal","header":395134233,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}},{"name":"from","type":{"kind":"simple","type":"address","optional":false}},{"name":"responseAddress","type":{"kind":"simple","type":"address","optional":true}},{"name":"forwardTonAmount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}},{"name":"forwardPayload","type":{"kind":"simple","type":"slice","optional":false,"format":"remainder"}}]},{"name":"TokenNotification","header":1935855772,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}},{"name":"from","type":{"kind":"simple","type":"address","optional":false}},{"name":"forwardPayload","type":{"kind":"simple","type":"slice","optional":false,"format":"remainder"}}]},{"name":"TokenBurn","header":1499400124,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}},{"name":"owner","type":{"kind":"simple","type":"address","optional":false}},{"name":"responseAddress","type":{"kind":"simple","type":"address","optional":true}}]},{"name":"TokenBurnNotification","header":2078119902,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}},{"name":"owner","type":{"kind":"simple","type":"address","optional":false}},{"name":"responseAddress","type":{"kind":"simple","type":"address","optional":true}}]},{"name":"TokenExcesses","header":3576854235,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}}]},{"name":"TokenUpdateContent","header":201882270,"fields":[{"name":"content","type":{"kind":"simple","type":"cell","optional":true}}]},{"name":"SetUserStatusMsg","header":2205882193,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}},{"name":"user","type":{"kind":"simple","type":"address","optional":false}},{"name":"message","type":{"kind":"simple","type":"string","optional":false}}]},{"name":"PositionState","header":null,"fields":[{"name":"collateral","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}},{"name":"debt","type":{"kind":"simple","type":"uint","optional":false,"format":64}}]},{"name":"PoolSettingsMsg","header":2701599449,"fields":[{"name":"liquidationRatio","type":{"kind":"simple","type":"uint","optional":false,"format":32}},{"name":"stabilityFeeRate","type":{"kind":"simple","type":"uint","optional":false,"format":32}},{"name":"closeFactorBps","type":{"kind":"simple","type":"uint","optional":false,"format":32}},{"name":"liquidatorIncentiveBps","type":{"kind":"simple","type":"uint","optional":false,"format":32}}]},{"name":"Deps","header":null,"fields":[{"name":"stablecoinMasterAddress","type":{"kind":"simple","type":"address","optional":false}},{"name":"positionsManagerAddress","type":{"kind":"simple","type":"address","optional":false}}]},{"name":"SetDeps","header":1512759282,"fields":[{"name":"stablecoinMasterAddress","type":{"kind":"simple","type":"address","optional":false}},{"name":"positionsManagerAddress","type":{"kind":"simple","type":"address","optional":false}}]},{"name":"UpdateTonPriceMsg","header":219265725,"fields":[{"name":"price","type":{"kind":"simple","type":"uint","optional":false,"format":32}}]},{"name":"DepositCollateralUserMessage","header":3295464374,"fields":[{"name":"user","type":{"kind":"simple","type":"address","optional":false}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}}]},{"name":"WithdrawStablecoinUserMessage","header":1885332798,"fields":[{"name":"user","type":{"kind":"simple","type":"address","optional":false}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}}]},{"name":"RepayStablecoinUserMessage","header":3726616898,"fields":[{"name":"user","type":{"kind":"simple","type":"address","optional":false}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}}]},{"name":"WithdrawCollateralUserMessage","header":3579692410,"fields":[{"name":"user","type":{"kind":"simple","type":"address","optional":false}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}}]},{"name":"WithdrawFeesMessage","header":960452073,"fields":[{"name":"to","type":{"kind":"simple","type":"address","optional":false}}]}],"receivers":[{"receiver":"internal","message":{"kind":"typed","type":"SetDeps"}},{"receiver":"internal","message":{"kind":"typed","type":"PoolSettingsMsg"}},{"receiver":"internal","message":{"kind":"typed","type":"UpdateTonPriceMsg"}},{"receiver":"internal","message":{"kind":"text","text":"updateDebtRate"}},{"receiver":"internal","message":{"kind":"typed","type":"WithdrawFeesMessage"}},{"receiver":"internal","message":{"kind":"typed","type":"DepositCollateralUserMessage"}},{"receiver":"internal","message":{"kind":"typed","type":"WithdrawStablecoinUserMessage"}},{"receiver":"internal","message":{"kind":"typed","type":"RepayStablecoinUserMessage"}},{"receiver":"internal","message":{"kind":"typed","type":"WithdrawCollateralUserMessage"}},{"receiver":"internal","message":{"kind":"typed","type":"WithdrawCollateralMessage"}},{"receiver":"internal","message":{"kind":"typed","type":"DecreaseTotalStableMessage"}},{"receiver":"internal","message":{"kind":"typed","type":"IncreaseTotalStableMessage"}},{"receiver":"internal","message":{"kind":"typed","type":"Deploy"}}],"getters":[{"name":"poolSettings","arguments":[],"returnType":{"kind":"simple","type":"PoolSettings","optional":false}},{"name":"tonPrice","arguments":[],"returnType":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"tonPriceWithSafetyMargin","arguments":[],"returnType":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"debtRate","arguments":[],"returnType":{"kind":"simple","type":"DebtRate","optional":false}},{"name":"stablecoinsIssued","arguments":[],"returnType":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"totalCollateralAmount","arguments":[],"returnType":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"getBalance","arguments":[],"returnType":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"getDeps","arguments":[],"returnType":{"kind":"simple","type":"Deps","optional":false}},{"name":"owner","arguments":[],"returnType":{"kind":"simple","type":"address","optional":false}}],"errors":{"2":{"message":"Stack undeflow"},"3":{"message":"Stack overflow"},"4":{"message":"Integer overflow"},"5":{"message":"Integer out of expected range"},"6":{"message":"Invalid opcode"},"7":{"message":"Type check error"},"8":{"message":"Cell overflow"},"9":{"message":"Cell underflow"},"10":{"message":"Dictionary error"},"13":{"message":"Out of gas error"},"32":{"message":"Method ID not found"},"34":{"message":"Action is invalid or not supported"},"37":{"message":"Not enough TON"},"38":{"message":"Not enough extra-currencies"},"128":{"message":"Null reference exception"},"129":{"message":"Invalid serialization prefix"},"130":{"message":"Invalid incoming message"},"131":{"message":"Constraints error"},"132":{"message":"Access denied"},"133":{"message":"Contract stopped"},"134":{"message":"Invalid argument"},"135":{"message":"Code of a contract was not found"},"136":{"message":"Invalid address"},"137":{"message":"Masterchain support is not enabled for this contract"},"4429":{"message":"Invalid sender"},"15032":{"message":"not from stablecoin master"},"22230":{"message":"Already set"},"23932":{"message":"not enough tons sent"},"31797":{"message":"debt less than repay amount"},"32589":{"message":"updateDebtAccumulatedRate:Invalid timestamp"},"41207":{"message":"invalid sender"},"43504":{"message":"position will not be healthy"},"53160":{"message":"not from positions manager"},"60380":{"message":"no fees collected"},"61910":{"message":"not from positionsManager"},"63577":{"message":"withdrawal amount more than position has"}},"interfaces":["org.ton.introspection.v0","org.ton.abi.ipfs.v0","org.ton.deploy.lazy.v0","org.ton.debug.v0","org.ton.chain.workchain.v0","org.ton.ownable"]}