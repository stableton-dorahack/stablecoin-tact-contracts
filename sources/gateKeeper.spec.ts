import { OpenedContract, toNano } from "ton";
import { ContractSystem, Logger, Tracker, Treasure } from "ton-emulator";
import { GateKeeperContract } from "./output/stableton_GateKeeperContract";

describe("GateKeeperContract", () => {
    let system: ContractSystem;
    let owner: Treasure;
    let user: Treasure;
    let positionsManager: Treasure;
    let stablecoinMaster: Treasure;
    let gateKeeperContract: OpenedContract<GateKeeperContract>;
    let track: Tracker;
    let logger: Logger;

    beforeAll(async () => {
        system = await ContractSystem.create();
        owner = system.treasure("owner");
        user = system.treasure("user");
        positionsManager = system.treasure("positionsManager");
        stablecoinMaster = system.treasure("stablecoinMaster");

        gateKeeperContract = system.open(await GateKeeperContract.fromInit());

        track = system.track(gateKeeperContract.address);
        logger = system.log(gateKeeperContract.address);
    });

    it("should deploy correctly", async () => {
        await gateKeeperContract.send(
            owner,
            { value: toNano(1) },
            {
                $$type: "SetDeps",
                stablecoinMasterAddress: stablecoinMaster.address,
                positionsManagerAddress: positionsManager.address,
            }
        );
        await system.run();

        expect(track.collect()).toMatchSnapshot();

        expect(await gateKeeperContract.getPoolSettings()).toEqual({
            $$type: "PoolSettings",
            closeFactorBps: 0n,
            liquidationRatio: 0n,
            liquidatorIncentiveBps: 0n,
            stabilityFeeRate: 0n,
        });

        const deps = await gateKeeperContract.getGetDeps();

        expect(deps.positionsManagerAddress.toString()).toEqual("EQCmw0-Nvnap6F8s-77yXEYbduFtd_bSsmDZu6eYT4Oy-LEr");
        expect(deps.stablecoinMasterAddress.toString()).toEqual("EQCDr0Gm-nivCD9KWPb6JkiDulGZ8eNb2Xm_HNi39oyNbKeD");
    });

    it("on pool settings msg should store new settings ", async () => {
        // Set pool settings by owner

        const now = Math.round(new Date().getTime() / 1000);
        await gateKeeperContract.send(
            owner,
            { value: toNano(1) },
            {
                $$type: "PoolSettingsMsg",
                liquidationRatio: 1200000000n,
                stabilityFeeRate: 2n,
                closeFactorBps: 2500n,
                liquidatorIncentiveBps: 10500n,
            }
        );
        await system.run();

        expect(track.collect()).toMatchSnapshot();

        expect(await gateKeeperContract.getPoolSettings()).toEqual({
            $$type: "PoolSettings",
            liquidationRatio: 1200000000n,
            stabilityFeeRate: 2n,
            closeFactorBps: 2500n,
            liquidatorIncentiveBps: 10500n,
        });
    });

    it("should set ton price", async () => {
        console.log("contract.getTonPrice()", await gateKeeperContract.getTonPrice());

        expect(await gateKeeperContract.getTonPrice()).toEqual(0n);
        expect(await gateKeeperContract.getTonPriceWithSafetyMargin()).toEqual(0n);

        // Set pool settings by owner
        await gateKeeperContract.send(
            owner,
            { value: toNano(1) },
            {
                $$type: "UpdateTonPriceMsg",
                price: 3200000000n,
            }
        );
        await system.run();
        expect(track.collect()).toMatchSnapshot();

        // Check updated ton price
        console.log("contract.getTonPrice()", await gateKeeperContract.getTonPrice());
        expect(await gateKeeperContract.getTonPrice()).toEqual(3200000000n);

        // liquidation ratio = 120%, means if collateral volume is less than debt*1.2 then liquidation possible
        console.log("tonPriceWithSafetyMargin", await gateKeeperContract.getTonPriceWithSafetyMargin());
        expect(await gateKeeperContract.getTonPriceWithSafetyMargin()).toEqual(2666666666n);
    });

    it("on deposit collateral should forward message to positionsManager and have msg.amount left on gatekeeper balance", async () => {
        console.log("gateKeeper balance before", await gateKeeperContract.getGetBalance());

        // send collateral 100 ton plus 1 ton for messaging
        await gateKeeperContract.send(
            user,
            { value: toNano(101) },
            {
                $$type: "DepositCollateralUserMessage",
                user: user.address,
                amount: toNano(100),
            }
        );
        await system.run();
        expect(track.collect()).toMatchSnapshot();

        // gatekeeper should store 100 ton collateral
        console.log("gateKeeper balance after", await gateKeeperContract.getGetBalance());
        expect(await gateKeeperContract.getGetBalance()).toBeGreaterThan(toNano(100));

        // console.warn(logger.collect());
    });

    it("on withdrawstablecoins should forward message to positionsManager with current settings and price", async () => {
        await gateKeeperContract.send(
            user,
            { value: toNano(1) },
            {
                $$type: "WithdrawStablecoinUserMessage",
                user: user.address,
                amount: toNano(100),
            }
        );
        await system.run();
        expect(track.collect()).toMatchSnapshot();
    });

    it("on increase total stable should add total stables count", async () => {
        const totalStablesIssuedBefore = await gateKeeperContract.getStablecoinsIssued();
        await gateKeeperContract.send(
            positionsManager,
            { value: toNano(1) },
            {
                $$type: "IncreaseTotalStableMessage",
                user: user.address,
                amount: toNano(100),
            }
        );
        await system.run();
        expect(track.collect()).toMatchSnapshot();

        const totalStablesIssuedAfter = await gateKeeperContract.getStablecoinsIssued();
        expect(totalStablesIssuedAfter - totalStablesIssuedBefore).toEqual(toNano(100));
    });

    // todo test for WithdrawFeesMessage
});
