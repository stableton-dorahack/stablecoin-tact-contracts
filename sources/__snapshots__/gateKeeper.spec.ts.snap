// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GateKeeperContract on deposit collateral should forward message to positionsManager and have msg.amount left on gatekeeper balance 1`] = `
[
  {
    "$seq": 3,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{C46CCBB6800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B2A2E90EDD001_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQA8J7zdze7xzxWC0UEFLI_ebJedZgQbQuay8b9oAkTSWQkO",
          "to": "kQBrMIVDTwvNGs9An7Y_GB-oXpPiM39ZoObutITWBvuOu2Dl",
          "type": "internal",
          "value": 101000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 14809n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "cell": "x{DDED1FE8800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B2A2E90EDD0008F0D180000000004000013880000520877359400C822BBBC93DE43555_}",
              "type": "cell",
            },
            "bounce": true,
            "from": "kQBrMIVDTwvNGs9An7Y_GB-oXpPiM39ZoObutITWBvuOu2Dl",
            "to": "kQCmw0-Nvnap6F8s-77yXEYbduFtd_bSsmDZu6eYT4Oy-Aqh",
            "type": "internal",
            "value": 978329000n,
          },
        ],
      },
    ],
  },
  {
    "$seq": 4,
    "events": [
      {
        "$type": "received-bounced",
        "message": {
          "body": {
            "type": "empty",
          },
          "bounce": false,
          "from": "kQCmw0-Nvnap6F8s-77yXEYbduFtd_bSsmDZu6eYT4Oy-Aqh",
          "to": "kQBrMIVDTwvNGs9An7Y_GB-oXpPiM39ZoObutITWBvuOu2Dl",
          "type": "internal",
          "value": 977329000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 8135n,
      },
    ],
  },
]
`;

exports[`GateKeeperContract on increase total stable should add total stables count 1`] = `
[
  {
    "$seq": 7,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{96096519800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B20000002E90EDD001_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQCmw0-Nvnap6F8s-77yXEYbduFtd_bSsmDZu6eYT4Oy-Aqh",
          "to": "kQBrMIVDTwvNGs9An7Y_GB-oXpPiM39ZoObutITWBvuOu2Dl",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 13164n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "type": "empty",
            },
            "bounce": false,
            "from": "kQBrMIVDTwvNGs9An7Y_GB-oXpPiM39ZoObutITWBvuOu2Dl",
            "to": "kQA8J7zdze7xzxWC0UEFLI_ebJedZgQbQuay8b9oAkTSWQkO",
            "type": "internal",
            "value": 985836000n,
          },
        ],
      },
    ],
  },
]
`;

exports[`GateKeeperContract on pool settings msg should store new settings  1`] = `
[
  {
    "$seq": 1,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{A10722D947868C0000000002000009C400002904}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQAI-3FJVc_ywSuY4vq0bYrzR7S4Och4y7bTU_i5yLOB3A6P",
          "to": "kQBrMIVDTwvNGs9An7Y_GB-oXpPiM39ZoObutITWBvuOu2Dl",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 15276n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "text": "pool settings updated",
              "type": "text",
            },
            "bounce": true,
            "from": "kQBrMIVDTwvNGs9An7Y_GB-oXpPiM39ZoObutITWBvuOu2Dl",
            "to": "kQAI-3FJVc_ywSuY4vq0bYrzR7S4Och4y7bTU_i5yLOB3A6P",
            "type": "internal",
            "value": 983424000n,
          },
        ],
      },
    ],
  },
]
`;

exports[`GateKeeperContract on repayStablecoinUserMessage should send enhanced message to positionsAddressManager 1`] = `
[
  {
    "$seq": 8,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{DE1FA942800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B2A2E90EDD001_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQA8J7zdze7xzxWC0UEFLI_ebJedZgQbQuay8b9oAkTSWQkO",
          "to": "kQBrMIVDTwvNGs9An7Y_GB-oXpPiM39ZoObutITWBvuOu2Dl",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 14286n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "cell": "x{6C3989BB800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B20000002E90EDD0008F0D180000000004000013880000520877359400C822BBBC93DE43555_}",
              "type": "cell",
            },
            "bounce": true,
            "from": "kQBrMIVDTwvNGs9An7Y_GB-oXpPiM39ZoObutITWBvuOu2Dl",
            "to": "kQCmw0-Nvnap6F8s-77yXEYbduFtd_bSsmDZu6eYT4Oy-Aqh",
            "type": "internal",
            "value": 984023000n,
          },
        ],
      },
    ],
  },
]
`;

exports[`GateKeeperContract on withdrawstablecoins should forward message to positionsManager with current settings and price 1`] = `
[
  {
    "$seq": 5,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{705FE53E800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B2A2E90EDD001_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQA8J7zdze7xzxWC0UEFLI_ebJedZgQbQuay8b9oAkTSWQkO",
          "to": "kQBrMIVDTwvNGs9An7Y_GB-oXpPiM39ZoObutITWBvuOu2Dl",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 14183n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "cell": "x{6C3989BB800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B20000002E90EDD0008F0D180000000004000013880000520877359400C822BBBC93DE43555_}",
              "type": "cell",
            },
            "bounce": true,
            "from": "kQBrMIVDTwvNGs9An7Y_GB-oXpPiM39ZoObutITWBvuOu2Dl",
            "to": "kQCmw0-Nvnap6F8s-77yXEYbduFtd_bSsmDZu6eYT4Oy-Aqh",
            "type": "internal",
            "value": 984126000n,
          },
        ],
      },
    ],
  },
  {
    "$seq": 6,
    "events": [
      {
        "$type": "received-bounced",
        "message": {
          "body": {
            "type": "empty",
          },
          "bounce": false,
          "from": "kQCmw0-Nvnap6F8s-77yXEYbduFtd_bSsmDZu6eYT4Oy-Aqh",
          "to": "kQBrMIVDTwvNGs9An7Y_GB-oXpPiM39ZoObutITWBvuOu2Dl",
          "type": "internal",
          "value": 983126000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 8135n,
      },
    ],
  },
]
`;

exports[`GateKeeperContract should deploy correctly 1`] = `
[
  {
    "$seq": 0,
    "events": [
      {
        "$type": "deploy",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{5A2ADFF2801075E834DF4F15E107E94B1EDF44C910774A333E3C6B7B2F37E39B16FED191AD90029B0D3E36F9DAA7A17CB3EEFBC971186DDB85B5DFDB4AC98366EE9E613E0ECBE2_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQAI-3FJVc_ywSuY4vq0bYrzR7S4Och4y7bTU_i5yLOB3A6P",
          "to": "kQBrMIVDTwvNGs9An7Y_GB-oXpPiM39ZoObutITWBvuOu2Dl",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 15943n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "text": "deps set",
              "type": "text",
            },
            "bounce": true,
            "from": "kQBrMIVDTwvNGs9An7Y_GB-oXpPiM39ZoObutITWBvuOu2Dl",
            "to": "kQAI-3FJVc_ywSuY4vq0bYrzR7S4Och4y7bTU_i5yLOB3A6P",
            "type": "internal",
            "value": 982861000n,
          },
        ],
      },
    ],
  },
]
`;

exports[`GateKeeperContract should set ton price 1`] = `
[
  {
    "$seq": 2,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{0D11BABDBEBC2000}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQAI-3FJVc_ywSuY4vq0bYrzR7S4Och4y7bTU_i5yLOB3A6P",
          "to": "kQBrMIVDTwvNGs9An7Y_GB-oXpPiM39ZoObutITWBvuOu2Dl",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 14736n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "text": "ton price updated",
              "type": "text",
            },
            "bounce": true,
            "from": "kQBrMIVDTwvNGs9An7Y_GB-oXpPiM39ZoObutITWBvuOu2Dl",
            "to": "kQAI-3FJVc_ywSuY4vq0bYrzR7S4Och4y7bTU_i5yLOB3A6P",
            "type": "internal",
            "value": 983996000n,
          },
        ],
      },
    ],
  },
]
`;
